### ISSUE TITLE
How to make the package type stable?

### ISSUE THREAD

**Issue by ali-ramadhan** [CliMA/CLIMAParameters.jl#44]
Not sure how much of a concern this is but for example, the mean sea-level pressure is always a `Float64`

https://github.com/CliMA/CLIMAParameters.jl/blob/f779db6fa86f0bfce9de4ee31757fed1fe8b75ed/src/Planet/planet_parameters.jl#L47

while the total solar irradiance is always an `Int`

https://github.com/CliMA/CLIMAParameters.jl/blob/f779db6fa86f0bfce9de4ee31757fed1fe8b75ed/src/Planet/planet_parameters.jl#L46

This might introduce lots of type conversions if you're running a `Float32` model, potentially impacting performance. Type conversions can completely kill performance during automatic differentiation and backpropagation.

So I'm guessing the onus is on the user, e.g. ClimateMachine.jl or Oceananigans.jl, to always perform the appropriate conversions?

One solution might be to add a parametric type (default is `Float64`) to the parameter sets that then takes care of all conversions:

```julia
abstract type AbstractParameterSet{FT} end

abstract type AbstractEarthParameterSet{FT} <: AbstractParameterSet{FT} end

AbstractEarthParameterSet(FT=Float64) = AbstractEarthParameterSet{FT}()

Planet.TSI(ps::AbstractEarthParameterSet{FT}) where FT  = FT(1362)
Planet.MSLP(ps::AbstractEarthParameterSet{FT}) where FT = FT(1.01325e5)
```

**Comment by charleskawczynski** [comment]
@simonbyrne and I had some discussions about this back when I was moving clima parameters from ClimateMachine.jl into CLIMAParameters.jl. We [used to](https://github.com/CliMA/ClimateMachine.jl/commit/6708ca669bda7468116b0846a216d2f627c1c586) have `FT` as a type parameter, but there were several considerations that lead us to remove it. The big one that comes to mind is that many of the tests/drivers/experiments, especially those with multiple float types, using clima parameters got really messy. Not having `FT` in the parameter type allows us to have a single `param_set` defined at the top of the file--and that is not the case if `FT` is included. The trade-off is that we need to locally enforce type-stability, which is something we need to be cognizant of anyway.

**Comment by glwagner** [comment]
> Not having FT in the parameter type allows us to have a single param_set defined at the top of the file--and that is not the case if FT is included.

I don't quite understand the reasoning --- can you explain @charleskawczynski ? Users have to specify a floating point type for their experiment at some point in their script. What's the downside to requiring this statement when building `parameters`, as opposed to some other place?

**Comment by charleskawczynski** [comment]
I found the [relevant PR](https://github.com/CliMA/ClimateMachine.jl/pull/893#issue-394360512) where we made this decision.

**Comment by charleskawczynski** [comment]
I remember really struggling with the decision to remove it--I did not want to, and that's why I looped in @simonbyrne.

Maybe things have changed since then that could make including `FT` easier now. Or maybe I was just fatigued by the [long process](https://github.com/CliMA/ClimateMachine.jl/issues/957) of extracting all of the parameters, and maybe it'd be easier to include `FT` now that they've all been extracted (from the original `PlanetParameters`).

It'd probably still be a significant effort, though. Personally, I don't feel too motivated as those PRs were some of the most difficult ones to decompose and get across the finish line.

**Comment by charleskawczynski** [comment]
Now that I'm thinking more about it, I think one of the reasons those PRs were difficult to merge was because of the way CI was configured at the time (all serially). Since many of the changes were across the test/experiment/driver files, it took seemingly forever to find out that the nth experiment failed. And it seemed like by the time I reached the last experiment, other PRs merged and I'd have to rebase (and re-apply changes to newly added code). It might be easier now with buildkite running most experiments in parallel, but we have even more experiments/tests now.

**Comment by ali-ramadhan** [comment]
Thanks for the explanation @charleskawczynski. Didn't realize a decision has already been made to remove `FT`.

I guess I saw my suggestion as non-breaking since the default `FT` is `Float64` so I think adding it wouldn't break anything in ClimateMachine.jl? Although I'm not familiar with ClimateMachine.jl so I can't say I suppose.

> The big one that comes to mind is that many of the tests/drivers/experiments, especially those with multiple float types, using clima parameters got really messy.

Do you have a link to an example of this? I'm not sure why there would be many float types. Seeing some of the messy code would help me understand some of the reasons you posted in PR https://github.com/CliMA/ClimateMachine.jl/pull/893.

**Comment by charleskawczynski** [comment]
> I guess I saw my suggestion as non-breaking since the default FT is Float64 so I think adding it wouldn't break anything in ClimateMachine.jl?

We mostly use `FT`-generic interfaces in ClimateMachine.jl. In several unit tests, we test multiple float types, which I believe is an [open issue](https://github.com/CliMA/ClimateMachine.jl/issues/273) in ClimateMachine.jl.

Here are a few places where we'd need `param_set = ...` in just one file:
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L48
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L104
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L458
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L642
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L727
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L975
 - https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L1094

With the existing approach, all we need is https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/runtests.jl#L18.

**Comment by glwagner** [comment]
> We mostly use FT-generic interfaces in ClimateMachine.jl. In several unit tests, we test multiple float types, which I believe is an open issue in ClimateMachine.jl.

@charleskawczynski are you confirming that the suggested feature is non-breaking, or saying that it _is_ breaking?

As far as I can tell, adding a default conversion to `Float64` would primarily have the effect of promoting the few `Int` parameters to `Float64` (leaving parameters already defined as `Float64` unchanged).

This would then be a feature that package develops could choose to take advantage of --- or not.

**Comment by charleskawczynski** [comment]
> @charleskawczynski are you confirming that the suggested feature is non-breaking, or saying that it is breaking?

If we added `FT` to a type parameter in CLIMAParameters.jl, we would need to make changes across ClimateMachine. So, in that sense, yes.

> This would then be a feature that package developers could choose to take advantage of --- or not.

I guess I'd need to see an example of your suggestion. Feel free to open a tracer bullet PR 🚀

**Comment by glwagner** [comment]
@charleskawczynski @ali-ramadhan has provided example code showing the implementation. The proposal is to add a type parameter `FT` to `CLIMAParameters` types. Parameters are converted to `FT` when the parameter functions are called; for example:

```julia
Planet.TSI(ps::AbstractEarthParameterSet{FT}) where FT  = FT(1362)
Planet.MSLP(ps::AbstractEarthParameterSet{FT}) where FT = FT(1.01325e5)
```

For packages that depend on `CLIMAParameters` like `ClimateMachine`, the effect of a default `FT=Float64` is hopefully clear. For the parameters that are currently `Float64` (like "MSLP"), there is no effect. For parameters that are currently `Int` (like "TSI") the addition of this default has the effect of promoting `Int` to `Float64`.

I don't think it's likely this is a breaking change. A type parameter `FT` with default `Float64` is an _optional_ feature. If a package like `ClimateMachine` doesn't want to use it because they would rather convert numbers internally, that options remains available. The only possibility of breaking behavior is if some part of `ClimateMachine` specifically assumes that `Int` parameters (like "TSI") are integers.

**Comment by charleskawczynski** [comment]
> A type parameter FT with default Float64 is an _optional_ feature.

How will users optionally create instances of `AbstractEarthParameterSet{FT}` subtypes?

```julia
julia> abstract type AbstractParameterSet{FT} end

julia> struct EarthParameterSet <: AbstractParameterSet{FT} end
ERROR: UndefVarError: FT not defined
Stacktrace:
 [1] top-level scope at REPL[2]:1
```

**Comment by glwagner** [comment]
Hmm I see. Something like

```julia
julia> abstract type AbstractTypedParameterSet{FT} end

julia> const AbstractParameterSet = AbstractTypedParameterSet{Float64}
AbstractTypedParameterSet{Float64}

julia> struct EarthParameterSet <: AbstractParameterSet end
```

might work.

**Comment by jakebolewski** [comment]
why does the abstract type need to be parameterized? any fall through is undefined (if there is no parameter one doesn't exist)

**Comment by charleskawczynski** [comment]
> Something like .... might work

That might work, but it may also cause confusion. You could turn it around and create your own instances and locally overload methods:

```julia
import CLIMAParameters
abstract type AbstractParameterSet end
struct MyEarthParameterSet{FT} <: AbstractParameterSet end

CLIMAParameters.Planet.planet_radius(::MyEarthParameterSet{FT}) where {FT} = FT(10^3)
```
You just wouldn't be able to impose the type restriction on methods in CLIMAParameters.

> I guess i don't know why there is an abstract earth and not just a concrete type earth

I think we wanted to be able to create custom Earth instances:
```julia
struct SmallEarth <: AbstractEarthParameterSet end
CLIMAParameters.Planet.planet_radius(::SmallEarth) = 10^3
```
so that it still inherits all other Earth parameters.

**Comment by charleskawczynski** [comment]
Closed by #79. (we no longer ship these types, we only define the values and users must define their own types).
